docs/guides/typescript.md

```diff
@@ -158,42 +158,30 @@ The typestates of a machine are specified as the 3rd generic type in `createMach
 **Example:**
 
 ```ts
-import { createMachine, interpret } from 'xstate';
+import {
+  CombineTypestates,
+  TypestateContext,
+  createMachine,
+  interpret
+} from 'xstate';
 
 interface User {
   name: string;
 }
 
-interface UserContext {
-  user?: User;
-  error?: string;
-}
-
 type UserEvent =
   | { type: 'FETCH'; id: string }
   | { type: 'RESOLVE'; user: User }
   | { type: 'REJECT'; error: string };
 
-type UserState =
-  | {
-      value: 'idle';
-      context: UserContext & {
-        user: undefined;
-        error: undefined;
-      };
-    }
-  | {
-      value: 'loading';
-      context: UserContext;
-    }
-  | {
-      value: 'success';
-      context: UserContext & { user: User; error: undefined };
-    }
-  | {
-      value: 'failure';
-      context: UserContext & { user: undefined; error: string };
-    };
+type UserState = CombineTypestates<
+  | { value: 'idle'; context: {} }
+  | { value: 'loading'; context: { user?: User; error?: string } }
+  | { value: 'success'; context: { user: User } }
+  | { value: 'failure'; context: { error: string } }
+>;
+
+type UserContext = TypestateContext<UserState>;
```

Comment:
how is this different from:
```js
type UserContext = TypestateContext<
  | { value: 'idle'; context: {} }
  | { value: 'loading'; context: { user?: User; error?: string } }
  | { value: 'success'; context: { user: User } }
  | { value: 'failure'; context: { error: string } }
>;
```

Reply:
TypestateContext requires a union where each context type has every key present, including the ones that are always undefined for the state value. So it's is actually identical to:

```ts
type UserContext = TypestateContext<
  | { value: 'idle';    context: { user?: undefined; error?: undefined } }
  | { value: 'loading'; context: { user?: User;      error?: string } }
  | { value: 'success'; context: { user:  User;      error?: undefined } }
  | { value: 'failure'; context: { user?: undefined; error: string } }
>;
```

The ideal API in my mind (maybe for v5?) would require only two generic params and no helpers:

```
type UserEvent =
  | { type: 'FETCH'; id: string }
  | { type: 'RESOLVE'; user: User }
  | { type: 'REJECT'; error: string };

type UserState =
  | { value: 'idle'; context: {} }
  | { value: 'loading'; context: { user?: User; error?: string } }
  | { value: 'success'; context: { user: User } }
  | { value: 'failure'; context: { error: string } };

const userMachine = Machine<UserState, UserEvent>({
      /* ... */
});
```

Reply:
> TypestateContext requires a union where each context type has every key present, including the ones that are always undefined for the state value. So it's is actually identical to:

How useful `CombineTypestates` is then? The main benefit I'm seeing (but I, of course, might be missing something) is the ability to merge/combine typestates to create a "typestate context". If that's the main appeal - why `TypestateContext` doesn't just "call" `CombineTypestates` internally (so it could be used like on the example I've given)?

 > The ideal API in my mind (maybe for v5?) would require only two generic params, and could use TypeScript 4.1 string template types

The idea of having this as an object with paths as the keys is appealing. Not sure if you have meant that - but I don't think `TContext` should be completely gone. It's by far easier to use than typestates and actually provides type safety, whereas using typestates can lead to runtime bugs because we can't guarantee that the declared types match the reality.

Reply:
Agreed, I think it should be possible to use the API with either a simple context type or a complex typestate union.

> It's by far easier to use than typestates and actually provides type safety, whereas using typestates can lead to runtime bugs because we can't guarantee that the declared types match the reality.

Maybe this can be examined for a future major version. I believe it should be possible to define the types of all typestates, events, and transitions so that an invalid state will not compile, but I'm not experienced enough with xstate yet to tell what API changes would need to be made.

Reply:
> How useful CombineTypestates is then? The main benefit I'm seeing (but I, of course, might be missing something) is the ability to merge/combine typestates to create a "typestate context". If that's the main appeal - why TypestateContext doesn't just "call" CombineTypestates internally (so it could be used like on the example I've given)?

It absolutely could, but there would still need to be some utility type to extract the typestate union from the context type so it could be passed as the third generic param to `createMachine`, to avoid altering the public API of `createMachine`.

My thinking here was that, in a future version, if the state schema could be strongly typed in a way that no transition allows an invalid context for a given state value, then the typestates would be more relevant for TypeScript users than the aggregate context shape.